### ISSUE TITLE
cost function used in discriminator of GAN

### ISSUE THREAD

**Issue by RiKiBom** [eriklindernoren/Keras-GAN#157]
d_loss_real = self.discriminator.train_on_batch(imgs, valid)
            d_loss_fake = self.discriminator.train_on_batch(gen_imgs, fake)
            d_loss = 0.5 * np.add(d_loss_real, d_loss_fake)

I think this part of code is wrong.
train_on_batch didn't use the sum of log(real) and log(1-fake).
It first used log(real) and then used log(1-fake).
Thank you.

**Comment by Revist** [comment]
I agree, I believe the code ought to be:

batch_imgs = np.concatenate(imgs, gen_imgs)
batch_labels = np.concatenate(valid, fake)
d_loss = self.discriminator.train_on_batch(batch_imgs, batch_labels)
d_loss *= 0.5

**Comment by fanstvn** [comment]
i concatenate the real images and fake images into one batch and feed it to the discriminator,
it is interesting that the generator convergence slower and more unstable.
![Figure_1](https://user-images.githubusercontent.com/30677462/69989824-70758000-157f-11ea-9d7e-bf38e012b5d6.png)
this is the loss curve of original code.
![Figure_2](https://user-images.githubusercontent.com/30677462/69989826-723f4380-157f-11ea-957c-98cfff98d243.png)

**Comment by fanstvn** [comment]
I concatenate real images and fake images to calculate the loss in original paper. The changes of training loss turns more sharply. It's probablely not a bug, it's a feature.😄
1/m in the paper averages the loss for a minibatch of m samples, which is divided by keras itself.
0.5 in the code averages the two losses(real and fake) to get the average loss for all sample of two batchs(real and fake).
It is not in the paper. But if you concatenate the real images and fake images with 'batch_size = m' both and train the model, the loss is actually averaged by 2*m.
